frame/democracy/src/lib.rs

```diff
@@ -1049,8 +1088,107 @@ pub mod pallet {
 					T::Slash::on_unbalanced(T::Currency::slash_reserved(&who, amount).0);
 				}
 			}
-
 			Self::deposit_event(Event::<T>::ProposalCanceled { prop_index });
+			Self::clear_metadata(MetadataOwner::Proposal(prop_index));
+			Ok(())
+		}
+
+		/// Set the metadata to the external proposal.
+		///
+		/// - `origin`: Must be an external origin and correspond to the proposal vote threshold,
+		///   `ExternalOrigin` for `SuperMajorityApprove`, `ExternalDefaultOrigin` for
+		///   `SuperMajorityAgainst` and `ExternalMajorityOrigin` for `SimpleMajority`.
+		/// - `hash`: The preimage hash of an on-chain stored preimage.
+		#[pallet::weight(T::WeightInfo::set_external_metadata())]
+		pub fn set_external_metadata(origin: OriginFor<T>, hash: PreimageHash) -> DispatchResult {
+			let (_, threshold) = <NextExternal<T>>::get().ok_or(Error::<T>::NoProposal)?;
+			match threshold {
+				VoteThreshold::SuperMajorityApprove => {
+					let _ = T::ExternalOrigin::ensure_origin(origin)?;
+				},
+				VoteThreshold::SuperMajorityAgainst => {
+					let _ = T::ExternalDefaultOrigin::ensure_origin(origin)?;
+				},
+				VoteThreshold::SimpleMajority => {
+					let _ = T::ExternalMajorityOrigin::ensure_origin(origin)?;
+				},
+			};
+			ensure!(T::Preimages::len(&hash).is_some(), Error::<T>::BadMetadata);
+			T::Preimages::request(&hash);
+			MetadataOf::<T>::insert(MetadataOwner::External, hash);
+			Self::deposit_event(Event::<T>::MetadataSet { owner: MetadataOwner::External, hash });
+			Ok(())
+		}
+
+		/// Set the metadata to the public proposal.
+		///
+		/// - `origin`: Must be `Signed`, and the creator of the proposal.
+		/// - `index`: The index of the proposal.
+		/// - `hash`: The preimage hash of an on-chain stored preimage.
+		#[pallet::weight(T::WeightInfo::set_proposal_metadata())]
+		pub fn set_proposal_metadata(
+			origin: OriginFor<T>,
+			index: PropIndex,
+			hash: PreimageHash,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let (_, _, proposer) = Self::proposal(index)?;
+			ensure!(proposer == who, Error::<T>::NoPermission);
+			ensure!(T::Preimages::len(&hash).is_some(), Error::<T>::BadMetadata);
```

Comment:
This can go before the 
`let (_, _, proposer) = Self::proposal(index)?;`, because it does not need to read storage. More efficient.

Reply:
In fact I misread it, `len` method name gave me the wrong idea. In which case just leave this as is.